src/mlpack/tests/main_tests/hmm_train_test.cpp

```diff
@@ -206,4 +301,104 @@ BOOST_AUTO_TEST_CASE(HMMTrainBatchModeTest)
+  int states = 2;
+
+  FileExists(inputObsFile1);
+  SetInputParam("input_file", std::move(inputObsFile1));
+  SetInputParam("type", std::move(type));
+  SetInputParam("states", states);
+
+  mlpackMain();
+
+  HMMModel h1 = *(CLI::GetParam<HMMModel*>("output_model"));
+
+  std::string inputObsFile2 = "obs3.csv";
+
+  CLI::GetSingleton().Parameters()["input_file"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["type"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["states"].wasPassed = false;
+
+  FileExists(inputObsFile2);
+  SetInputParam("input_file", std::move(inputObsFile2));
+  SetInputParam("input_model", CLI::GetParam<HMMModel*>("output_model"));
+
+  mlpackMain();
+
+  HMMModel h2 = *(CLI::GetParam<HMMModel*>("output_model"));
+
+  BOOST_REQUIRE(!ApproximatelyEqual(h1, h2, 1e-04));
+}
+
+// Attempt to retrain but increase states the second time round
+BOOST_AUTO_TEST_CASE(HMMTrainRetrainTest2)
+{
+  // Provide no labels file
+  std::string inputObsFile1 = "obs1.csv";
+  std::string type = "discrete";
+  int states = 2;
+
+  SetInputParam("input_file", std::move(inputObsFile1));
+  SetInputParam("type", std::move(type));
+  SetInputParam("states", states);
+
+  mlpackMain();
+
+  HMMModel h1 = *(CLI::GetParam<HMMModel*>("output_model"));
+
+  std::string inputObsFile2 = "obs3.csv";
+  std::string inputLabFile2 = "lab1_corrupt.csv";
+
+  SetInputParam("input_file", std::move(inputObsFile2));
+  // Provide a labels file with more states than initially specified
+  SetInputParam("labels_file", std::move(inputLabFile2));
+  SetInputParam("input_model", CLI::GetParam<HMMModel*>("output_model"));
+
+  CLI::GetSingleton().Parameters()["type"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["states"].wasPassed = false;
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+// Attempt to retrain but change the emission distribution type
+BOOST_AUTO_TEST_CASE(HMMTrainRetrainTest3)
+{
+  // Provide no labels file
+  std::string inputObsFile1 = "obs1.csv";
+  std::string type = "discrete";
+  int states = 2;
+
+  SetInputParam("input_file", std::move(inputObsFile1));
+  SetInputParam("type", std::move(type));
+  SetInputParam("states", states);
+
+  mlpackMain();
+
+  HMMModel h1 = *(CLI::GetParam<HMMModel*>("output_model"));
+
+  std::string inputObsFile2 = "obs2.csv";
+  std::string inputLabFile2 = "lab2.csv";
+  type = "gaussian";
+
+  SetInputParam("input_file", std::move(inputObsFile2));
+  // Provide a labels file with more states than initially specified
+  SetInputParam("labels_file", std::move(inputLabFile2));
+  SetInputParam("type", std::move(type));
+  SetInputParam("input_model", CLI::GetParam<HMMModel*>("output_model"));
+
+  CLI::GetSingleton().Parameters()["type"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["states"].wasPassed = false;
+
+  HMMModel h2 = *(CLI::GetParam<HMMModel*>("output_model"));
```

Comment:
I think you are forgetting a call to `mlpackMain()`. :)  It may also be worthwhile to add a comment pointing out that when the emission type is changed between runs, a warning will be printed but it's not a fatal error.  (I think that is true...)